judge/contest_format/ioi16.py

```diff
@@ -0,0 +1,98 @@
+from django.db import connection
+from django.utils.translation import gettext_lazy
+
+from judge.contest_format.ioi import IOIContestFormat
+from judge.contest_format.registry import register_contest_format
+from judge.timezone import from_database_time
+
+
+@register_contest_format('ioi16')
+class IOI16ContestFormat(IOIContestFormat):
+    name = gettext_lazy('IOI16')
+    config_defaults = {'cumtime': False}
+    '''
+        cumtime: Specify True if time penalties are to be computed. Defaults to False.
+    '''
+
+    def __init__(self, contest, config):
+        super(IOI16ContestFormat, self).__init__(contest, config)
+
+    def update_participation(self, participation):
+        cumtime = 0
+        score = 0
+        format_data = {}
+
+        with connection.cursor() as cursor:
+            cursor.execute('''
+                SELECT *
+                FROM (
+                         SELECT q.prob,
+                                q.date,
+                                q.batch_points
+                         FROM (
+                                  SELECT cp.id          as `prob`,
+                                         sub.id         as `subid`,
+                                         sub.date       as `date`,
+                                         tc.points      as `points`,
+                                         tc.batch       as `batch`,
+                                         Min(tc.points) as `batch_points`
+                                  FROM judge_contestproblem cp
+                                           INNER JOIN
+                                       judge_contestsubmission cs
+                                       ON (cs.problem_id = cp.id AND cs.participation_id = %s)
+                                           LEFT OUTER JOIN
+                                       judge_submission sub ON (sub.id = cs.submission_id)
+                                           INNER JOIN judge_submissiontestcase tc
+                                                      ON sub.id = tc.submission_id
+                                  GROUP BY cp.id, tc.batch, sub.id
```

Comment:
I think you should filter out only finished submissions, i.e. `status = 'D'`. I think this will prevent abortions and other failures from inflating batch points.

Reply:
If, for example, a submission gets aborted right after batch 1 was scored and actually got full score on that batch, should this submission be ignored?

Reply:
I don't think there is a general method to tell if a batch is actually fully graded (because the site doesn't know how many cases are in a batch, or how many batches there are), so I think we should just ignore the submission entirely.

Reply:
To add onto this: that would make it consistent with the other contest formats, too. Right now, if you time your abort request right, it can happen when the judge sends the last test case packet and when it sends its "I'm done grading" packet, so you can end up with an aborted solution that has all of the test cases present (but of course, the site can't know this).

In general, I don't think making aborted solution scoring particularly robust is worth the effort.